gobblin-runtime/src/main/java/gobblin/runtime/instance_driver/DefaultGobblinInstanceDriverImpl.java

```diff
@@ -0,0 +1,173 @@
+  protected final JobCatalog _jobCatalog;
+  protected final JobSpecScheduler _jobScheduler;
+  protected final JobExecutionLauncher _jobLauncher;
+  protected JobSpecListener _jobSpecListener;
+
+  public DefaultGobblinInstanceDriverImpl(JobCatalog jobCatalog, JobSpecScheduler jobScheduler,
+      JobExecutionLauncher jobLauncher, Optional<Logger> log) {
+    Preconditions.checkNotNull(jobCatalog);
+    Preconditions.checkNotNull(jobScheduler);
+    Preconditions.checkNotNull(jobLauncher);
+
+    _jobCatalog = jobCatalog;
+    _jobScheduler = jobScheduler;
+    _jobLauncher = jobLauncher;
+    _log = log.isPresent() ? log.get() : LoggerFactory.getLogger(getClass());
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public JobCatalog getJobCatalog() {
+    return _jobCatalog;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public JobSpecScheduler getJobScheduler() {
+    return _jobScheduler;
+  }
+
+  /** {@inheritDoc} */
+  @Override
+  public JobExecutionLauncher getJobLauncher() {
+    return _jobLauncher;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    _jobSpecListener = new JobSpecListener();
+    _jobCatalog.addListener(_jobSpecListener);
+  }
+
+  @Override
+  protected void shutDown() throws Exception {
+    if (null != _jobSpecListener) {
+      _jobCatalog.removeListener(_jobSpecListener);
+    }
+  }
+
+  class ExecutionStateListener extends DefaultJobExecutionStateListenerImpl {
+
+    public ExecutionStateListener() {
+      super(DefaultGobblinInstanceDriverImpl.this._log);
+    }
+
+    @Override
+    public void onStatusChange(JobExecutionState state, RunningState previousStatus, RunningState newStatus) {
+      // TODO Auto-generated method stub
+
+    }
+
+    @Override
+    public void onStageTransition(JobExecutionState state, String previousStage, String newStage) {
+      // TODO Auto-generated method stub
+
+    }
+
+    @Override
+    public void onMetadataChange(JobExecutionState state, String key, Object oldValue, Object newValue) {
+      // TODO Auto-generated method stub
+
+    }
+
+  }
+
+  /** Keeps track of a job execution */
+  class JobExecutionListener extends DefaultJobExecutionStateListenerImpl {
+
+    public JobExecutionListener() {
+      super(LoggerFactory.getLogger(DefaultGobblinInstanceDriverImpl.this._log.getName() +
+                                  "_jobExecutionListener"));
+    }
+
+    @Override public String toString() {
+      return _log.get().getName();
+    }
+
+  }
+
+  /** The runnable invoked by the Job scheduler */
+  class JobSpecRunnable implements Runnable {
+    private final JobSpec _jobSpec;
+
+    public JobSpecRunnable(JobSpec jobSpec) {
+      _jobSpec = jobSpec;
+    }
+
+    @Override
+    public void run() {
+       _jobLauncher.launchJob(_jobSpec, new JobExecutionListener());
+    }
+  }
+
+  /** Listens to changes in the Job catalog and schedules/un-schedules jobs. */
+  class JobSpecListener extends DefaultJobCatalogListenerImpl {
+
```

Comment:
DefaultJobCatalogListenerImpl is meant to provide token implementation for all callbacks. Other classes can extend from it and override only the methods they care about. If we use delegation pattern, we'll have to always implement all the methods.